lib/pmdtester/builders/diff_builder.rb

```diff
@@ -0,0 +1,202 @@
+require 'nokogiri'
+
+module PmdTester
+  class DiffBuilder
+
+    # The schema of pmd xml report refer to
+    # http://pmd.sourceforge.net/report_2_0_0.xsd
+
+    def build(base_report, patch_report)
+      base_doc = Nokogiri::XML(File.read(base_report))
+      patch_doc = Nokogiri::XML(File.read(patch_report))
+
+      violations_diffs = build_violation_diffs(base_doc, patch_doc)
+      error_diffs = build_error_diffs(base_doc, patch_doc)
+
+      [violations_diffs, error_diffs]
+    end
+
+    def build_violation_diffs(base_doc, patch_doc)
+
+      # key:String => value:Violation Array
+      diffs = Hash.new
+
+      base_doc.xpath('//file').each do |file|
+        filename, violations = get_violations_in_file(file, 'base')
+        diffs.store(filename, violations)
+      end
+
+      patch_doc.xpath('//file').each do|file|
+        filename, violations = get_violations_in_file(file, 'patch')
+
+       if diffs.has_key?(filename)
+         diffs[filename] = remove_duplicate(diffs[filename], violations)
+         diffs.delete(filename) if diffs[filename].empty?
+       else
+         diffs.store filename, [violations, 'patch']
+       end
+      end
+      diffs
+    end
+
+    def get_violations_in_file(file, branch)
+
+      # The shcema of 'file' node:
+      #  <xs:complexType name="file">
+      #    <xs:sequence>
+      #      <xs:element name="violation" type="violation" minOccurs="1" maxOccurs="unbounded" />
+      #      </xs:sequence>
+      #    <xs:attribute name="name" type="xs:string" use="required"/>
+      #  </xs:complexType>
+
+      filename = file['name']
+      violations = []
+      file.xpath('violation').each do |violation|
+        violations.push(PmdViolation.new(violation, branch))
+      end
+      [filename, violations]
+    end
+
+    def remove_duplicate(base_violations, patch_violations)
+      i, j = 0, 0
+      diff = []
+      while i < base_violations.size && j < patch_violations.size
```

Comment:
Do we maybe need to sort the violations first? I'm not sure, in which order the different rules will be executed and end up in the report, so potentially, the base and patch violations could contain exactly the same violations, but in total different order.

Have you looked at https://docs.ruby-lang.org/en/2.0.0/Hash.html? I know from java, there are good support methods to deal with Hashes (or Sets/HashSet). Since you already use concrete objects (PmdViolation), it might be easier to do the following: Implement `hash` and `eql?` in PmdViolation. Then you need to figure out the duplicates: Dups = Base intersect Patch . The end result would be "Base union Patch" minus "Duplicates". That should give you the violations, that are either in base or in patch, but not in both....
Or in Java:

```
Set<PmdViolation> result = new HashSet<>();
Set<PmdViolation> base = ...;
Set<PmdViolation> patch = ...;

result.addAll(base);
result.addAll(patch);

Set<PmdViolation> intersection = new HashSet<>(base);
intersection.retainAll(patch);

result.removeAll(intersection);
```

(Caveat: I didn't try this out....)